File: Leetcode/MinimumTimetoMakeRopeColorful.rs
// https://leetcode.com/problems/minimum-time-to-make-rope-colorful

struct Solution {}

impl Solution {
    pub fn min_cost(colors: String, needed_time: Vec<i32>) -> i32 {
        let colors = colors.as_bytes();
        let mut cur_ind = 0;
        let mut cur = 0;
        let mut ans = 0;

        while cur_ind < colors.len() {
            let mut sum = 0;
            let mut mx = i32::MIN;
            cur = colors[cur_ind];
            while cur_ind < colors.len() && cur == colors[cur_ind] {
                sum += needed_time[cur_ind];
                mx = mx.max(needed_time[cur_ind]);
                cur_ind += 1;
            }

            ans += sum - mx;
        }

        ans
    }
}

File: Leetcode/MinimumTimetoMakeRopeColorful.cpp
// https://leetcode.com/problems/minimum-time-to-make-rope-colorful

class Solution {
   public:
    int minCost(string colors, vector<int>& neededTime) {
        auto ans = 0;
        auto start_it = colors.begin();
        auto end_it = colors.begin();

        while (end_it != colors.cend()) {
            while (end_it != colors.cend() && *start_it == *end_it) {
                ++end_it;
            }

            auto sum = 0;
            auto ma = neededTime[start_it - colors.cbegin()];
            while (start_it != end_it) {
                const auto time = neededTime[start_it - colors.cbegin()];
                sum += time;
                if (ma < time) {
                    ma = time;
                }
                ++start_it;
            }

            ans += sum - ma;
        }

        return ans;
    }
};